File: 6. Week 6 - Dashboard/solutions/r/visualization.R
# Visualize Running Dashboard using ggplot2 and patchwork

if (!require(ggplot2)) install.packages("ggplot2")
if (!require(readr)) install.packages("readr")
if (!require(patchwork)) install.packages("patchwork") # For dashboard layout
library(ggplot2)
library(readr)
library(patchwork)

# --- 1. Load Data ---
df_bar <- read_csv("../../data/distance_per_month.csv", show_col_types = FALSE)
df_scatter <- read_csv("../../data/distance_vs_pace.csv", show_col_types = FALSE)
df_line <- read_csv("../../data/trailing_365_distance.csv", show_col_types = FALSE)
df_pie <- read_csv("../../data/time_of_day_counts.csv", show_col_types = FALSE)

# --- 2. Create Individual Plots ---

# Bar Chart
p1 <- ggplot(df_bar, aes(x = Month, y = TotalDistance)) +
  geom_col(fill = "rebeccapurple") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Monthly Distance", x = NULL, y = "km")

# Scatter Plot
p2 <- ggplot(df_scatter, aes(x = Distance_km, y = AveragePace_min_km, color = AveragePace_min_km)) +
  geom_point(alpha = 0.6) +
  scale_color_gradient(low = "plum", high = "rebeccapurple") +
  theme_light() +
  theme(legend.position = "none") +
  labs(title = "Distance vs Pace", x = "km", y = "min/km")

# Line Chart
p3 <- ggplot(df_line, aes(x = Date, y = TotalDistance)) +
  geom_line(color = "rebeccapurple") +
  scale_x_date(date_labels = "%Y", date_breaks = "1 year") +
  theme_minimal() +
  labs(title = "Trailing 365d", x = NULL, y = "km")

# Pie Chart (Approximation via Bar for dashboard simplicity or Polar)
# Polar plots can be tricky in patchwork layout sizing, but let's try.
p4 <- ggplot(df_pie, aes(x = "", y = Count, fill = Category)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar("y", start = 0) +
  scale_fill_manual(values = c("AM" = "#9575CD", "PM" = "#673AB7", "Both" = "#D1C4E9")) +
  theme_void() +
  labs(title = "Time of Day")

# --- 3. Assemble Dashboard ---
# Using patchwork syntax: (p1 | p2) / (p3 | p4)
dashboard <- (p1 + p2) / (p3 + p4) +
  plot_annotation(title = "Running Analysis Dashboard",
                  theme = theme(plot.title = element_text(size = 20, hjust = 0.5)))

ggsave("dashboard.png", plot = dashboard, width = 12, height = 10)
print("Dashboard saved to dashboard.png")

File: 6. Week 6 - Dashboard/solutions/python/visualization.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

def visualize():
    # Setup Figure
    fig = plt.figure(figsize=(20, 12))
    
    # Grid Layout: 2x2
    gs = fig.add_gridspec(2, 2)
    
    # --- Chart 1: Bar Chart (Distance per Month) ---
    ax1 = fig.add_subplot(gs[0, 0])
    df1 = pd.read_csv('../../data/distance_per_month.csv')
    ax1.bar(df1['Month'], df1['TotalDistance'], color='rebeccapurple')
    ax1.set_title('Distance per Month')
    ax1.tick_params(axis='x', rotation=45)
    
    # --- Chart 2: Scatter Plot (Distance vs Pace) ---
    ax2 = fig.add_subplot(gs[0, 1])
    df2 = pd.read_csv('../../data/distance_vs_pace.csv')
    sc = ax2.scatter(df2['Distance_km'], df2['AveragePace_min_km'], c=df2['AveragePace_min_km'], cmap='Purples')
    ax2.set_title('Distance vs Pace')
    ax2.set_xlabel('Distance (km)')
    ax2.set_ylabel('Pace (min/km)')
    
    # --- Chart 3: Line Chart (Trailing 365) ---
    ax3 = fig.add_subplot(gs[1, 0])
    df3 = pd.read_csv('../../data/trailing_365_distance.csv')
    df3['Date'] = pd.to_datetime(df3['Date'])
    ax3.plot(df3['Date'], df3['TotalDistance'], color='rebeccapurple')
    ax3.set_title('Trailing 365-Day Distance')
    ax3.tick_params(axis='x', rotation=45)
    
    # --- Chart 4: Pie Chart (Time of Day) ---
    ax4 = fig.add_subplot(gs[1, 1])
    df4 = pd.read_csv('../../data/time_of_day_counts.csv')
    ax4.pie(df4['Count'], labels=df4['Category'], autopct='%1.1f%%', colors=['#D1C4E9', '#9575CD', '#673AB7', '#311B92'])
    ax4.set_title('Time of Day')
    
    # Final Layout
    plt.tight_layout()
    plt.suptitle('Running Dashboard', fontsize=24, y=1.02)
    plt.savefig('dashboard.png', bbox_inches='tight')
    print("Dashboard saved to dashboard.png")
    plt.show()

if __name__ == "__main__":
    visualize()
